middleware/openidconnect.go

```diff
@@ -0,0 +1,240 @@
+package middleware
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	oidc "github.com/coreos/go-oidc"
+	"github.com/owncloud/ocis-pkg/log"
+	ocisoidc "github.com/owncloud/ocis-pkg/oidc"
+	"golang.org/x/oauth2"
+)
+
+// OIDCOptions for the OpenID Connect middleware
+type OIDCOptions struct {
+	// Logger to use for logging, if not set will default to panic level logging
+	Logger *log.Logger
+	// Endpoint is the OpenID Connect provider URL
+	Endpoint string
+	// Realm to use in the WWW-Authenticate header, defaults to Endpoint
+	Realm string
+	// Audience to use when checking jwt based tokens
+	Audience string
+	// SigningAlgs to use when verifying jwt signatures, defaults to "RS256" & "PS256"
+	SigningAlgs []string
+	// ClientId to use as username for basic auth against the introspection endpoint
+	ClientID string
+	// ClientSecret to use as password for basic auth against the introspection endpoint
+	ClientSecret string
+	// Insecure can be used to disable http certificate checks
+	Insecure bool
+	// SkipCheck can be used to further reduce security. Fix that!
+	SkipChecks bool
+}
+
+// OpenIDConnect provides a middleware to check access secured by a static token.
+func OpenIDConnect(opt OIDCOptions) func(http.Handler) http.Handler {
+
+	var l log.Logger
+	if opt.Logger == nil {
```

Comment:
If you are using the same options pattern we are using already everywhere within the ocis repos you can define defaults much cleaner.

Reply:
ok, so I move the options to individual functions inside `pkg/ocis/options.go` and implement a `newOptions(o Option...)` that fills the options object that is private to the middleware. right? at least somewhat?

Reply:
@tboerger currently the middlewares all take explicit arguments ... that would be quite a list in this case ... I'd prefer to have a `func OpenIDConnect(opts ...Option) func(http.Handler) http.Handler {` signature than 9 parameters

Reply:
ok. @tboerger Does this make sense? I am asking because all other middlewares take explicit parameters ...

Reply:
`func OpenIDConnect(opts ...Option) func(http.Handler) http.Handler {` is what I was refering to. That's the scheme we are using for all the ocis repos.

Reply:
but for middlewares like `Static` we have [`func Static(root string, fs http.FileSystem) func(http.Handler) http.Handler {`](https://github.com/owncloud/ocis-pkg/blob/master/middleware/static.go#L10). They don't follow that pattern because ... it's only two parameters? When do we start using the _options pattern_ ... if that is a thing ... 3 parameters? 4? when the number starts to suck?

Just trying to find something we can note down as a coding guideline ... somewhere...

Reply:
Indeed I have not done that for these 2 cases of the middlewares because there are only 2 options. Feel free to add options for all of them to keep it consistent with the other repos.